src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/TrafficSeo/Meta/MetaSettingsFormDataProvider.php

```diff
@@ -0,0 +1,223 @@
+        $this->setUpUrlDataConfiguration = $setUpUrlDataConfiguration;
+        $this->shopUrlsDataConfiguration = $shopUrlsDataConfiguration;
+        $this->urlSchemaDataConfiguration = $urlSchemaDataConfiguration;
+        $this->routeValidator = $routeValidator;
+        $this->translator = $translator;
+        $this->validate = $validate;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getData()
+    {
+        return [
+            'set_up_urls' => $this->setUpUrlDataConfiguration->getConfiguration(),
+            'shop_urls' => $this->shopUrlsDataConfiguration->getConfiguration(),
+            'url_schema' => $this->urlSchemaDataConfiguration->getConfiguration(),
+        ];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setData(array $data)
+    {
+        $errors = $this->validateData($data);
+
+        if (!empty($errors)) {
+            return $errors;
+        }
+
+        return array_merge(
+            $this->setUpUrlDataConfiguration->updateConfiguration($data['set_up_urls']),
+            $this->shopUrlsDataConfiguration->updateConfiguration($data['shop_urls']),
+            $this->urlSchemaDataConfiguration->updateConfiguration($data['url_schema'])
+        );
+    }
+
+    /**
+     * Implements custom validation for configuration form.
+     *
+     * @param array $data
+     *
+     * @return array - if array is not empty then error strings are returned
+     *
+     * @throws PrestaShopException
+     */
+    private function validateData(array $data)
+    {
+        $urlSchemaErrors = $this->validateUrlSchema($data['url_schema']);
+        $shopUrlErrors = $this->validateShopUrl($data['shop_urls']);
+
+        return array_merge(
+            $urlSchemaErrors,
+            $shopUrlErrors
+        );
+    }
+
+    /**
+     * Validates if configuration matches route pattern and if route has mandatory fields.
+     *
+     * @param array $configuration
+     *
+     * @return array
+     *
+     * @throws PrestaShopException
+     */
+    private function validateUrlSchema(array $configuration)
+    {
+        $patternErrors = [];
+        $requiredFieldErrors = [];
+        foreach ($configuration as $routeId => $rule) {
+            if (!$this->routeValidator->isRoutePattern($rule)) {
+                $patternErrors[] = $this->translator->trans(
+                  'The route %routeRule% is not valid',
+                  [
+                      '%routeRule%' => htmlspecialchars($rule),
+                  ],
+                  'Admin.Shopparameters.Feature'
+                );
+            }
+
+            $missingKeywords = $this->routeValidator->doesRouteContainsRequiredKeywords($routeId, $rule);
+
+            if (!empty($missingKeywords)) {
+                foreach ($missingKeywords as $keyword) {
+                    $requiredFieldErrors[] = $this->translator->trans(
+                        'Keyword "{%keyword%}" required for route "%routeName%" (rule: "%routeRule%")',
```

Comment:
FYI, the following wording is available in _Admin.Shopparameters.Notification_: There are several available keywords for each route listed below; note that keywords with * are required!

Reply:
We won't forget it but we need this pull request to be merged before the freeze or @colinegin will kill me at some point :fearful: